# MeRIPseqPipe

**MeRIP-seq analysis pipeline arranged multiple alignment tools, peakCalling tools, Merge Peaks\' methods and methylation analysis methods.**.

[![Build Status](https://travis-ci.com/nf-core/meripseqpipe.svg?branch=master)](https://travis-ci.com/nf-core/meripseqpipe)
[![Nextflow](https://img.shields.io/badge/nextflow-%E2%89%A519.04.0-brightgreen.svg)](https://www.nextflow.io/)

[![install with bioconda](https://img.shields.io/badge/install%20with-bioconda-brightgreen.svg)](http://bioconda.github.io/)
[![Docker](https://img.shields.io/docker/automated/kingzhuky/meripseqpipe.svg)](https://hub.docker.com/r/kingzhuky/meripseqpipe)

## Introduction

Here, we present MeRIPseqPipe, an integrated analysis pipeline for MeRIP-seq data based on Nextflow. It integrates ten main functional modules including data preprocessing, quality control, read mapping, peak calling, peak merging, motif searching, peak annotation, differential methylation analysis, differential expression analysis, and data visualization, which covers the basic analysis of MeRIP-seq data. All the analysis modules are generated by [Nextflow](https://www.nextflow.io/), and all the third-party tools are encapsulated in the [Docker](https://www.docker.com/resources/what-container) container.

## Quick Start

1. install [`nextflow`](https://nf-co.re/usage/installation)

2. pull [`docker`](https://docs.docker.com/engine/installation/) image from dockerhub: [`kingzhuky/meripseqpipe:dev`](https://hub.docker.com/r/kingzhuky/meripseqpipe)

3. cloning this repository

    ```bash
    git clone https://github.com/canceromics/MeRIPseqPipe.git
    nextflow run /path/to/MeRIPseqPipe --help
    ```

4. test it on a minimal dataset with a single command

    ```bash
    nextflow run path/to/meripseqpipe -profile test,docker
    ```

5. Start running your own analysis!

    ```bash
    nextflow run path/to/meripseqpipe -profile docker --designfile designfile.tsv --comparefile compare.txt -resume --aligners star --fasta hg38_genome.fa --gtf gencode.v25.annotation.gtf --rRNA_fasta hg38_rRNA.fasta --outdir path/to/results --skip_createbedgraph --peakMerged_mode rank --star_index hg38/starindex --skip_meyer --skip_matk --methylation_analysis_mode Wilcox-test
    ```

See [usage docs](docs/usage.md) for all of the available options when running the pipeline.

## Documentation

The MeRIPseqPipe documentation is split into the following files:

1. [Usage](docs/usage.md)
    * [Parameter Documentation](docs/parameter_docs.md)
    * An overview of how the pipeline works, how to run it and a description of all of the different command-line flags.
    * let us know if you need more customization!
2. [Output](docs/output.md)
    * An overview of the different results produced by the pipeline
3. [Troubleshooting](docs/troubleshooting.md)

## Credits

MeRIPseqPipe was originally written by Xiaoqiong Bao, Kaiyu Zhu.

## Contributions and Support

If you would like to contribute to this pipeline, please see the [contributing guidelines](.github/CONTRIBUTING.md).

## Citation
